### PR TITLE
8276985: AArch64: [vectorapi] Backend support of VectorMaskToLongNode

### PR DESCRIPTION
**Pull request by theRealELiu**
The lack of codegen for VectorMaskToLong results in a regression on
AArch64 for VectorMask.laneIsSet, which relies on the intrinsification
of VectorMask.toLong after JDK-8273949.

This patch implements bitmask extraction on AArch64 for NEON and SVE by
using scalar instructions, which is equivalent to the PMOVMSK
instructions on X86. The performance of VectorMask.laneIsSet improves
about 10x for NEON and 2x~4x for SVE on my test machines.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8276985](https://bugs.openjdk.java.net/browse/JDK-8276985): AArch64: [vectorapi] Backend support of VectorMaskToLongNode

### Reviewers
 * [Ningsheng Jian](https://openjdk.java.net/census#njian) (@nsjian - Committer) ⚠️ Review applies to 0c84cbe4ab4fc45316880b1445df83dc01db7669
 * [Andrew Haley](https://openjdk.java.net/census#aph) (@theRealAph - **Reviewer**)

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/6585/head:pull/6585` \
`$ git checkout pull/6585`

Update a local copy of the PR: \
`$ git checkout pull/6585` \
`$ git pull https://git.openjdk.java.net/jdk pull/6585/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 6585`

View PR using the GUI difftool: \
`$ git pr show -t 6585`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/6585.diff">https://git.openjdk.java.net/jdk/pull/6585.diff</a>

</details>

### CODE REVIEW

**Comment by theRealELiu** [comment]
Could anyone help to review this patch?

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 [hunk 1]
```diff
@@ -2481,3 +2481,36 @@ instruct vmask_lasttrue16B(iRegINoSp dst, vecX src) %{
...
+    // Input "src" is a vector of boolean represented as
+    // bytes with 0x00/0x01 as element values.
+
+    __ fmovd(as_Register($dst$$reg), as_FloatRegister($src$$reg));
+    __ bytemask_compress(as_Register($dst$$reg));
+  %}
+  ins_pipe(pipe_slow);
+%}
+
+instruct vmask_tolong16B(iRegLNoSp dst, vecX src, iRegL tmp) %{
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2499]: "iRegL tmp" should be "iRegLNoSp tmp" or use rscratch directly?

File: src/hotspot/cpu/aarch64/aarch64_sve_ad.m4 [hunk 2]
```diff
@@ -3174,4 +3174,21 @@ instruct vmask_lasttrue_partial(iRegINoSp dst, pReg src, pReg ptmp, rFlagsReg cr
     __ sve_vmask_lasttrue($dst$$Register, bt, as_PRegister($ptmp$$reg), as_PRegister($ptmp$$reg));
   %}
   ins_pipe(pipe_slow);
+%}
+
+instruct vmask_tolong(iRegLNoSp dst, pReg src, vReg vtmp1, vReg vtmp2, pRegGov pgtmp, iRegL tmp, rFlagsReg cr) %{
```
Comment (nsjian) [inline src/hotspot/cpu/aarch64/aarch64_sve_ad.m4:3179]: And this iRegL

File: src/hotspot/cpu/aarch64/aarch64_neon_ad.m4 (hunk 1 above)
Comment (theRealAph) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4:2499]: Yes, good catch. This kind of bug is horrible to detect.

Comment (theRealELiu) [reply src/hotspot/cpu/aarch64/aarch64_neon_ad.m4]: Thanks. It's fixed now.

**Review by nsjian — APPROVED** [review]

**Comment by theRealELiu** [comment]
@theRealAph May I ask if this patch could catch up the RDP1?

**Comment by theRealAph** [comment]
> @theRealAph May I ask if this patch could catch up the RDP1?

I think we'd have to move the JEP to targeted. It's one of the unfortunate consequences of JEPs.

**Review by theRealAph — APPROVED** [review]

**Comment by theRealELiu** [comment]
Thanks for your review. @theRealAph @nsjian

**Pull request closed without merging**